identity/registry/registry_contract.go

```diff
@@ -104,22 +199,176 @@ func (registry *contractRegistry) SubscribeToRegistrationEvent(id identity.Ident
 		sink := make(chan *bindings.RegistryRegisteredIdentity)
 		subscription, err := registry.filterer.WatchRegisteredIdentity(filterOps, sink, userIdentities, accountantIdentities)
 		if err != nil {
-			registrationEvent <- Cancelled
-			log.Error().Err(err).Msg("")
+			registry.publisher.Publish(RegistrationEventTopic, RegistrationEventPayload{
+				ID:     identity,
+				Status: RegistrationError,
+			})
+			log.Error().Err(err).Msg("could not register to identity events")
+
+			updateErr := registry.storage.Store(StoredRegistrationStatus{
+				Identity:           identity,
+				RegistrationStatus: RegistrationError,
+			})
+			if updateErr != nil {
+				log.Error().Err(updateErr).Msg("could not store registration status")
+			}
 			return
 		}
 		defer subscription.Unsubscribe()
 
+		// TODO: maybe add appropriate timeout?
 		select {
-		case <-stopLoop:
-			registrationEvent <- Cancelled
 		case <-sink:
-			registrationEvent <- Registered
+			log.Info().Msgf("received registration event for %v", identity)
+			s, err := registry.storage.Get(identity)
+			if err != nil {
+				log.Error().Err(err).Msg("could not store registration status")
+			}
+
+			status := RegisteredConsumer
+			if s.RegistrationStatus == Promoting {
+				status = RegisteredProvider
+			}
+
+			log.Debug().Msgf("sending registration success event for %v", identity)
+			registry.publisher.Publish(RegistrationEventTopic, RegistrationEventPayload{
+				ID:     identity,
+				Status: status,
+			})
+
+			err = registry.storage.Store(StoredRegistrationStatus{
+				Identity:           identity,
+				RegistrationStatus: status,
+			})
+			if err != nil {
+				log.Error().Err(err).Msg("could not store registration status")
+			}
 		case err := <-subscription.Err():
+			registry.publisher.Publish(RegistrationEventTopic, RegistrationEventPayload{
+				ID:     identity,
+				Status: RegistrationError,
+			})
 			if err != nil {
 				log.Error().Err(err).Msg("Subscription error")
 			}
+			updateErr := registry.storage.Store(StoredRegistrationStatus{
+				Identity:           identity,
+				RegistrationStatus: RegistrationError,
+			})
+			if updateErr != nil {
+				log.Error().Err(updateErr).Msg("could not store registration status")
+			}
 		}
 	}()
 	return
 }
+
+func (registry *contractRegistry) handleStop() {
+	registry.once.Do(func() {
+		log.Info().Msg("Stopping registry...")
+		close(registry.stop)
+	})
+}
+
+func (registry *contractRegistry) handleStart() {
+	log.Info().Msg("starting registry...")
```

Comment:
I have stopped marking all logs to be capitalized in the PR.
I have created an issue that will help to keep it consistent in the future: https://github.com/mysteriumnetwork/node/issues/1414